sympy/matrices/common.py

```diff
@@ -2540,3 +2554,85 @@ def classof(A, B):
             return A.__class__
 
     raise TypeError("Incompatible classes %s, %s" % (A.__class__, B.__class__))
+
+
+def sumprodsimp(a, b, simplify=True):
+    """Sum-of-products with optional intermediate product simplification
+    targeted at the kind of blowup that occurs during summation of products.
+    Intended to reduce expression blowup during matrix multiplication or other
+    similar operations.
+
+    Parameters
+    ==========
+
+    a, b : iterable
+        These will be multiplied then summed together either normally or
+        using simplification on the intermediate products and cancelling at
+        the end according to the 'simplify' flag. The elements must already be
+        sympyfied and the sequences need not be of the same length, the shorter
+        will be used.
+
+    simplify : bool
+        When set intermediate and final simplification will be used, not set
+        will indicate a normal sum of produces.
+    """
+
+    expr = S.Zero
+    itra = iter(a)
+    itrb = iter(b)
+
+    # simple non-simplified sum of products
+    if not simplify:
+        try:
+            expr = next(itra)*next(itrb)
+
+            for a in itra:
+                expr += a*next(itrb)
+
+        except StopIteration:
+            pass
+
+        return expr
+
+    # part 1, the expanded summation
+    try:
+        prod    = next(itra)*next(itrb)
+        _expand = getattr(prod, 'expand', None)
+        expr    = _expand(power_base=False, power_exp=False, log=False, \
+                multinomial=False, basic=False) if _expand else prod
+
+        for a in itra:
+            prod     = a*next(itrb)
+            _expand  = getattr(prod, 'expand', None)
+            expr    += _expand(power_base=False, power_exp=False, log=False, \
+                    multinomial=False, basic=False) if _expand else prod
+
+    except StopIteration:
```

Comment:
I'm sure we can use some kind of for loop rather than catching StopIteration

Reply:
Should it not be an error if `len(b) < len(a)`?

The normal way to achieve what you have in both loops above is to use zip.

Reply:
I wrote a more straight-forward version of this code here: https://github.com/sympy/sympy/pull/17247#discussion_r317373309

The only difference is since you've changed the signature you'd need:
```python
    terms = (ai * bi for ai, bi in zip(a, b))
```

Reply:
That's why the code I just linked uses reduce which is what would be on the line below the code you've just quoted.

Reply:
I don't understand why reduce can't be used. This is the whole point of reduce:
```julia
In [1]: terms = [[2], [3, 4]]

In [2]: sum(terms)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-1a5d48817e89> in <module>()
----> 1 sum(terms)

TypeError: unsupported operand type(s) for +: 'int' and 'list'

In [3]: reduce(lambda a, b: a+b, terms)
Out[3]: [2, 3, 4]
```

Reply:
Can you show (using lists as I did above) why reduce doesn't work?

Reply:
I'm adding the lists as an example of something that you can add but you cannot add 0 to (just like the matrix elements in a block matrix).

If you want a dot-product then it would be:
```julia
In [3]: from functools import reduce                                                                                              

In [4]: def dot(a, b): 
   ...:     terms = (ai * bi for ai, bi in zip(a, b)) # zip truncates for unequal lengths... 
   ...:     return reduce(lambda a, b: a+b, terms) 
   ...:                                                                                                                           

In [5]: dot([1, 2], [3, 4])                                                                                                       
Out[5]: 11

In [6]: dot([1], [3])                                                                                                             
Out[6]: 3

In [7]: dot([], [])                                                                                                               
---------------------------------------------------------------------------
TypeError 
```
You could catch that TypeError if it is necessary. I would avoid it though since you have the same problem that reduce has: what should you return (if you need to consider block matrices etc)? Handling the matrix product with zero dimensions should be dealt with outside of the dot function I think.